crates/wasi-common/src/fdentry.rs

```diff
@@ -101,48 +43,38 @@ impl fmt::Debug for Descriptor {
 }
 
 impl Descriptor {
+    pub(crate) fn try_clone(&self) -> io::Result<Descriptor> {
+        match self {
+            Descriptor::OsHandle(file) => file.try_clone().map(|f| OsHandle::from(f).into()),
+            Descriptor::VirtualFile(virt) => virt.try_clone().map(Descriptor::VirtualFile),
+            Descriptor::Stdin => Ok(Descriptor::Stdin),
+            Descriptor::Stdout => Ok(Descriptor::Stdout),
+            Descriptor::Stderr => Ok(Descriptor::Stderr),
+        }
+    }
+
     /// Return a reference to the `OsHandle` or `VirtualFile` treating it as an
     /// actual file/dir, and allowing operations which require an actual file and
     /// not just a stream or socket file descriptor.
-    pub(crate) fn as_file<'descriptor>(&'descriptor self) -> Result<Handle<'descriptor>> {
+    pub(crate) fn as_file<'descriptor>(&'descriptor self) -> Result<&'descriptor Descriptor> {
```

Comment:
The explicit lifetime name here isn't necessary anymore, if this looks like a good path I'll remove it.

Reply:
I think I like where this is going, but I'm a bit concerned about the naming here. It seems that now `as_file` only checks whether the underlying entity *can* be dereferenced as either OS file/dir or virtual file, and wrapping `self` in `Result`. It seems what we're missing here is some sort of a common interface between the `OsHandle` and your now deleted `Handle` that would allow us to definitively separate `OsHandle + VirtualFile` from streams, etc. What do you guys think @iximeow @sunfishcode?

Reply:
I agree. An earlier version of this PR tried to do just such a thing by making `Descriptor` an enum of `Handle` or `Stream`, where `as_file` would return `Option<Handle>`, but eventually ran into some complications due to `OsHandleRef` and fell back to this. I think this is back to not using `OsHandleRef`, so maybe it's time to revisit that.

Reply:
Cool! I guess we can refactor this in a subsequent PR :-)